src/connector/src/sink/encoder/avro.rs

```diff
@@ -0,0 +1,897 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use apache_avro::schema::Schema as AvroSchema;
+use apache_avro::types::{Record, Value};
+use apache_avro::Writer;
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::Row;
+use risingwave_common::types::{DataType, DatumRef, ScalarRefImpl, StructType};
+use risingwave_common::util::iter_util::{ZipEqDebug, ZipEqFast};
+
+use super::{FieldEncodeError, Result as SinkResult, RowEncoder, SerTo};
+
+type Result<T> = std::result::Result<T, FieldEncodeError>;
+
+pub struct AvroEncoder<'a> {
```

Comment:
The encoder is taking a reference to a schema stored outside the encoder. IIUC, in the future when we port the encoder in this PR to formatter, the reference will refer to a schema field stored in the formatter. If so, since it's unsafe to implement self reference field in struct, we are unlikely to store the encoder as a field in the formatter, because otherwise the encoder field will refer to the schema field in the same formatter struct. If we are not able to store the encoder as a field in the formatter, we will have to recreate a new encoder in every `write_chunk` call, which may introduce extra unnecessary cost (in the current avro encoder, we will repeatedly call `validate_fields` ). Therefore I think we should make the encoder static without any reference.

For the proto encoder, it's not hard to make it static, because its output is static. However, for the current avro encoder, it seems more complex because its output is also referring to the avro schema. If we want to make it static, we either use GAT on the `Output` of the `RowEncoder` trait, or we just make the output type of avro schema as the static `Vec<u8>`. Personally I prefer making the output type as `Vec<u8>` so that we don't have to introduce too much complexity in the row encoder trait definition, and currently I didn't see any potential use case that will need the intermediate (avro record, avro schema).

Reply:
Made them lifetime-free just now. `proto` is easy because it has `Arc` inside. We just need to `Arc` for `avro` as well.

I also considered outputting `Vec<u8>` directly as well. The extra `SerTo` step exists from the beginning mainly for:
* Some existing callers of `JsonEncoder` want `Map<String, Value>` rather than `String`.
* Also for json, depending on the sink, some may want `String` while others may want `Vec<u8>`

Neither of them may not be necessary after some refactoring. I was just trying to avoid touching too much code at once.